File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/bus/service/BusApplicantService.java
package b1nd.dodam.domain.rds.bus.service;

import b1nd.dodam.domain.rds.bus.entity.Bus;
import b1nd.dodam.domain.rds.bus.entity.BusApplicant;
import b1nd.dodam.domain.rds.bus.exception.BusAlreadyAppliedPositionException;
import b1nd.dodam.domain.rds.bus.exception.BusApplicantNotFoundException;
import b1nd.dodam.domain.rds.bus.repository.BusApplicantRepository;
import b1nd.dodam.domain.rds.member.entity.Student;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class BusApplicantService {
    private final BusApplicantRepository busApplicantRepository;

    public List<BusApplicant> getByBus(Bus bus) {
        return busApplicantRepository.findByBus(bus);
    }

    public void existsBySeatsAndBus(Bus bus, int seat) {
        if (busApplicantRepository.findByBusAndSeat(bus, seat) != null)
            throw new BusAlreadyAppliedPositionException();
    }

    public void deleteByBus(long busId) {
        busApplicantRepository.deleteByBus_Id(busId);
    }

    public BusApplicant getByStudent(Student student) {
        return busApplicantRepository.findByStudent(student)
            .orElseThrow(BusApplicantNotFoundException::new);
    }

    public BusApplicant getByStudentOrNull(Student student) {
        return busApplicantRepository.findByStudent(student).orElse(null);
    }

    public void save(BusApplicant busApplicant) {
        busApplicantRepository.save(busApplicant);
    }
}

File: services/service-bus/src/main/kotlin/com/b1nd/dodamdodam/bus/domain/bus/service/BusApplicantService.kt
package com.b1nd.dodamdodam.bus.domain.bus.service

import com.b1nd.dodamdodam.bus.domain.bus.entity.BusApplicantEntity
import com.b1nd.dodamdodam.bus.domain.bus.entity.BusEntity
import com.b1nd.dodamdodam.bus.domain.bus.exception.BusAlreadyAppliedPositionException
import com.b1nd.dodamdodam.bus.domain.bus.exception.BusApplicantNotFoundException
import com.b1nd.dodamdodam.bus.domain.bus.repository.BusApplicantRepository
import org.springframework.stereotype.Service

@Service
class BusApplicantService(
    private val busApplicantRepository: BusApplicantRepository
) {
    fun getByBus(bus: BusEntity): List<BusApplicantEntity> {
        return busApplicantRepository.findAllByBus(bus)
    }

    fun checkSeatAvailable(bus: BusEntity, seat: Int) {
        if (busApplicantRepository.findByBusAndSeat(bus, seat) != null) {
            throw BusAlreadyAppliedPositionException()
        }
    }

    fun deleteByBusId(busId: Long) {
        busApplicantRepository.deleteAllByBusId(busId)
    }

    fun getByUserId(userId: String): BusApplicantEntity {
        return busApplicantRepository.findByUserId(userId)
            ?: throw BusApplicantNotFoundException()
    }

    fun getByUserIdOrNull(userId: String): BusApplicantEntity? {
        return busApplicantRepository.findByUserId(userId)
    }

    fun create(applicant: BusApplicantEntity) {
        busApplicantRepository.save(applicant)
    }

    fun createAll(applicants: List<BusApplicantEntity>) {
        busApplicantRepository.saveAll(applicants)
    }
}
